Panic in Cache.MatchingClusterQueues: ClusterQueue left in cache with nil NamespaceSelector when its cohort has a cycle

**What happened**:

kueue-controller-manager crashes, and can crash repeatedly, after a user creates a Cohort parent cycle plus a ClusterQueue in it. 

The cycle leaves a half-initialized ClusterQueue in the scheduler cache with a nil `NamespaceSelector`, and the next Namespace update anywhere in the cluster panics the controller with a nil pointer dereference. Confirmed live on released v0.19.0 and observed in CI on main.

### Mechanism

The cache adds a ClusterQueue in two steps: insert (`pkg/cache/scheduler/cache.go:207`), then populate fields. The cohort cycle makes population fail (`pkg/cache/scheduler/clusterqueue.go:163-164`) before `NamespaceSelector` is set (line 177), and the failure is only logged (`pkg/controller/core/clusterqueue_controller.go:336-338`): the half-built entry stays in the cache. Every Namespace update then calls `NamespaceSelector.Matches(...)` on all cached CQs (`cache.go:1120`); the nil selector panics.

### Evidence (live repro, v0.19.0 on kind)

The webhook accepts the cycle. All objects create cleanly. Controller log:

```
"caller":"core/clusterqueue_controller.go:337","msg":"Failed to add clusterQueue to cache","clusterQueue":{"name":"cycle-cq"},"error":"cohort has a cycle"
```

One namespace label change later:

```
panic: runtime error: invalid memory address or nil pointer dereference
sigs.k8s.io/kueue/pkg/cache/scheduler.(*Cache).MatchingClusterQueues
	/workspace/pkg/cache/scheduler/cache.go:1120
sigs.k8s.io/kueue/pkg/controller/core.(*cqNamespaceHandler).Update
	/workspace/pkg/controller/core/clusterqueue_controller.go:446
```

Pod restarts with exit code 2; on repeat it entered `CrashLoopBackOff`. Same stack in CI on main: [baseline-1-34](https://gcsweb.k8s.io/gcs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_kueue/13295/pull-kueue-test-e2e-baseline-main-1-34/2080271060514639872/artifacts/run-test-e2e-baseline-1.34.8/) (panic recurred 3×), [baseline-1-36](https://gcsweb.k8s.io/gcs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_kueue/13295/pull-kueue-test-e2e-baseline-main-1-36/2080271064096575488/artifacts/run-test-e2e-baseline-1.36.1/).

Recurrence is event-order dependent: after a restart the rebuilt cache may miss the cycle (CQ processed before both parent edges) and behave normally, but recreating the ClusterQueue while the cycle exists reliably re-poisons it.

### Secondary finding

While poisoned, the ClusterQueue's Active condition is self-contradictory on v0.19.0:

```json
{"type":"Active","status":"False","reason":"Ready","message":"Can admit new workloads"}
```

On main the same CQ reported `Active: True` (see #13295). Either way, status hides that the scheduler cache rejected the CQ.

**What you expected to happen**:

A cohort cycle is a config error (one typo'd `parentName` closes a cycle) and should degrade gracefully: reject or deactivate the ClusterQueue with an accurate condition, keep the controller running.

Candidate fixes: remove the CQ from the hierarchy manager when the add fails, and/or reorder `updateClusterQueue` so unconditional field assignments happen before the fallible cohort-tree update. A nil guard in `MatchingClusterQueues` alone would just hide the half-built entry.

**How to reproduce it (as minimally and precisely as possible)**:

1. Apply:

    ```yaml
    apiVersion: kueue.x-k8s.io/v1beta2
    kind: Cohort
    metadata:
      name: cohort-x
    spec:
      parentName: cohort-y
    ---
    apiVersion: kueue.x-k8s.io/v1beta2
    kind: Cohort
    metadata:
      name: cohort-y
    spec:
      parentName: cohort-x
    ---
    apiVersion: kueue.x-k8s.io/v1beta2
    kind: ResourceFlavor
    metadata:
      name: default-rf
    ---
    apiVersion: kueue.x-k8s.io/v1beta2
    kind: ClusterQueue
    metadata:
      name: cycle-cq
    spec:
      cohortName: cohort-x
      resourceGroups:
      - coveredResources: [cpu]
        flavors:
        - name: default-rf
          resources:
          - name: cpu
            nominalQuota: 1
    ```

2. `kubectl label namespace default repro-trigger=1 --overwrite` — controller panics and restarts.

3. If no panic after a controller restart (event ordering): delete and re-apply `cycle-cq`, then label a namespace again.

Regression unit test: in `pkg/cache/scheduler/cache_test.go`, add a CQ under a cohort cycle, then call `MatchingClusterQueues`.

**Anything else we need to know?**:

Surfaced by the hierarchical-cohort e2e tests in #13295 (cycle scenario from #13237), which currently work around the crash with a create-retry loop; this fix should land first so that workaround can be dropped.

Disclosure: investigated and drafted with AI assistance; the repro ran on a live kind cluster and all log evidence is verbatim.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.34.0 (kind, local); v1.34.8 / v1.36.1 (CI)
- Kueue version (use `git describe --tags --dirty --always`): v0.19.0 (local); `v0.20.0-devel-18-g2ffcff98b` (main, CI)
- Cloud provider or hardware configuration: kind
- OS / Kernel: n/a
- Install tools: `kubectl apply --server-side -f https://github.com/kubernetes-sigs/kueue/releases/download/v0.19.0/manifests.yaml`
- Others: n/a

---
> AI assistance used for creating this report

--- a/pkg/cache/scheduler/cache.go
+++ b/pkg/cache/scheduler/cache.go
@@ -192,6 +192,7 @@ func (c *Cache) newClusterQueue(log logr.Logger, cq *kueue.ClusterQueue) (*clust
 		Name:                kueue.ClusterQueueReference(cq.Name),
 		Workloads:           make(map[workload.Reference]*workload.Info),
 		WorkloadsNotReady:   sets.New[workload.Reference](),
+		NamespaceSelector:   labels.Nothing(),
 		localQueues:         make(map[queue.LocalQueueReference]*LocalQueue),
 		podsReadyTracking:   c.podsReadyTracking,
 		workloadInfoOptions: c.workloadInfoOptions,

--- a/pkg/cache/scheduler/cache_test.go
+++ b/pkg/cache/scheduler/cache_test.go
@@ -2823,6 +2823,98 @@ func TestMatchingClusterQueues(t *testing.T) {
 	}
 }
 
+// TestMatchingClusterQueuesAfterFailedUpdate covers ClusterQueues that stay in the
+// hierarchy manager after updateClusterQueue returns an error, since Namespace events
+// keep matching against them.
+func TestMatchingClusterQueuesAfterFailedUpdate(t *testing.T) {
+	teamSelector := func(team string) *metav1.LabelSelector {
+		return &metav1.LabelSelector{MatchLabels: map[string]string{"team": team}}
+	}
+	addCohortCycle := func(t *testing.T, cache *Cache) {
+		t.Helper()
+		if err := cache.AddOrUpdateCohort(utiltestingapi.MakeCohort("cycle-a").Parent("cycle-b").Obj()); err != nil {
+			t.Fatal(err)
+		}
+		if err := cache.AddOrUpdateCohort(utiltestingapi.MakeCohort("cycle-b").Parent("cycle-a").Obj()); err == nil {
+			t.Fatal("Expected failure when cycle")
+		}
+	}
+
+	cases := map[string]struct {
+		setup       func(*testing.T, *Cache)
+		wantMatch   []map[string]string
+		wantNoMatch []map[string]string
+	}{
+		"add into cyclic cohort applies the configured selector": {
+			setup: func(t *testing.T, cache *Cache) {
+				ctx, _ := utiltesting.ContextWithLog(t)
+				addCohortCycle(t, cache)
+				cq := utiltestingapi.MakeClusterQueue("cq").
+					Cohort("cycle-a").
+					NamespaceSelector(teamSelector("eng")).
+					Obj()
+				if err := cache.AddClusterQueue(ctx, cq); err == nil {
+					t.Fatal("Expected failure when adding cq to cohort with cycle")
+				}
+			},
+			wantMatch:   []map[string]string{{"team": "eng"}},
+			wantNoMatch: []map[string]string{{"team": "ops"}},
+		},
+		"update into cyclic cohort applies the new selector": {
+			setup: func(t *testing.T, cache *Cache) {
+				ctx, log := utiltesting.ContextWithLog(t)
+				cq := utiltestingapi.MakeClusterQueue("cq").
+					NamespaceSelector(teamSelector("eng")).
+					Obj()
+				if err := cache.AddClusterQueue(ctx, cq); err != nil {
+					t.Fatal(err)
+				}
+				addCohortCycle(t, cache)
+				updated := utiltestingapi.MakeClusterQueue("cq").
+					Cohort("cycle-a").
+					NamespaceSelector(teamSelector("ops")).
+					Obj()
+				if err := cache.UpdateClusterQueue(log, updated); err == nil {
+					t.Fatal("Expected failure when updating cq to cohort with cycle")
+				}
+			},
+			wantMatch:   []map[string]string{{"team": "ops"}},
+			wantNoMatch: []map[string]string{{"team": "eng"}},
+		},
+		"add with an unparsable selector matches no Namespace": {
+			setup: func(t *testing.T, cache *Cache) {
+				ctx, _ := utiltesting.ContextWithLog(t)
+				cq := utiltestingapi.MakeClusterQueue("cq").
+					NamespaceSelector(teamSelector("not a valid label value")).
+					Obj()
+				if err := cache.AddClusterQueue(ctx, cq); err == nil {
+					t.Fatal("Expected failure when adding cq with an unparsable selector")
+				}
+			},
+			wantNoMatch: []map[string]string{nil, {"team": "eng"}},
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			cache := New(utiltesting.NewFakeClient())
+			tc.setup(t, cache)
+
+			for _, nsLabels := range tc.wantMatch {
+				want := sets.New[kueue.ClusterQueueReference]("cq")
+				if diff := cmp.Diff(want, cache.MatchingClusterQueues(nsLabels)); diff != "" {
+					t.Errorf("MatchingClusterQueues(%v) returned unexpected ClusterQueues (-want,+got):\n%s", nsLabels, diff)
+				}
+			}
+			for _, nsLabels := range tc.wantNoMatch {
+				want := sets.New[kueue.ClusterQueueReference]()
+				if diff := cmp.Diff(want, cache.MatchingClusterQueues(nsLabels)); diff != "" {
+					t.Errorf("MatchingClusterQueues(%v) returned unexpected ClusterQueues (-want,+got):\n%s", nsLabels, diff)
+				}
+			}
+		})
+	}
+}
+
 // TestWaitForPodsReadyCancelled ensures that the WaitForPodsReady call does not block when the context is closed.
 func TestWaitForPodsReadyCancelled(t *testing.T) {
 	now := time.Now().Truncate(time.Second)

--- a/pkg/cache/scheduler/clusterqueue.go
+++ b/pkg/cache/scheduler/clusterqueue.go
@@ -155,6 +155,12 @@ func (c *clusterQueue) updateClusterQueue(
 	admissionChecks map[kueue.AdmissionCheckReference]AdmissionCheck,
 	oldParent *cohort,
 ) error {
+	nsSelector, err := metav1.LabelSelectorAsSelector(in.Spec.NamespaceSelector)
+	if err != nil {
+		return err
+	}
+	c.NamespaceSelector = nsSelector
+
 	if c.updateQuotasAndResourceGroups(in.Spec.ResourceGroups) || oldParent != c.Parent() {
 		if oldParent != nil && oldParent != c.Parent() {
 			updateCohortTreeResourcesIfNoCycle(oldParent)
@@ -170,12 +176,6 @@ func (c *clusterQueue) updateClusterQueue(
 			updateClusterQueueResourceNode(c)
 		}
 	}
-
-	nsSelector, err := metav1.LabelSelectorAsSelector(in.Spec.NamespaceSelector)
-	if err != nil {
-		return err
-	}
-	c.NamespaceSelector = nsSelector
 
 	c.isStopped = ptr.Deref(in.Spec.StopPolicy, kueue.None) != kueue.None
 

